config/cmake/version_describe.cmake

```diff
@@ -0,0 +1,52 @@
+#
+# Copyright (c) 2017-2019, Battelle Memorial Institute; Lawrence Livermore National
+# Security, LLC; Alliance for Sustainable Energy, LLC.  See the top-level NOTICE for additional details. All rights reserved. 
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+find_package(Git)
+
+# this function is to get a nicely formatted build string for later reference in a human readable format that has some utility for automation.  
+# the resulting format would be empty if the build matches a tag exactly, and otherwise <branch_name>-<git hash>[-dirty]  if dirty is present there are uncommitted changes in the repo
+# called like git_version_describe(${PROJECT_SOURCE_DIR} result)  then result will contain the description string
+
+function(git_version_describe source_path result)
+  message(STATUS "Source ${source_path}")
+  if (GIT_FOUND)
+	execute_process(COMMAND ${GIT_EXECUTABLE} -C ${source_path} --work-tree ${source_path} describe --tags OUTPUT_VARIABLE TAG_DESCRIPTION ERROR_VARIABLE TAGGING_ERROR OUTPUT_STRIP_TRAILING_WHITESPACE)
+	if (TAGGING_ERROR)
+		execute_process(COMMAND ${GIT_EXECUTABLE} -C ${source_path} --work-tree ${source_path} describe OUTPUT_VARIABLE TAG_DESCRIPTION OUTPUT_STRIP_TRAILING_WHITESPACE)
+	endif()
+	string(LENGTH ${TAG_DESCRIPTION} tag_length)
+	if (tag_length GREATER 16)
+	    string(FIND ${TAG_DESCRIPTION} "-" last_dash_loc REVERSE)
+		string(SUBSTRING ${TAG_DESCRIPTION} ${last_dash_loc} -1 hash_string)
+		execute_process(COMMAND ${GIT_EXECUTABLE} -C ${source_path} --work-tree ${source_path} describe --all --tags --dirty OUTPUT_VARIABLE TAG_DESCRIPTION2 OUTPUT_STRIP_TRAILING_WHITESPACE)
+		string(FIND ${TAG_DESCRIPTION2} ${hash_string} hash_loc)
+		if (hash_loc LESS 0)
+			string(FIND ${TAG_DESCRIPTION2} "-dirty" dirty_loc)
+			if (dirty_loc GREATER 0)
+				string(SUBSTRING ${TAG_DESCRIPTION2} 0 ${dirty_loc} tag_desc)
+				string(APPEND tag_desc ${hash_string} "-dirty")
+			else()
+				set(tag_desc ${TAG_DESCRIPTION2})
+				string(APPEND tag_desc ${hash_string})
+			endif()
+		else()
+			set(tag_desc ${TAG_DESCRIPTION2})
+		endif()
+		string(FIND ${tag_desc} "heads/" headloc)
+		if (headloc EQUAL 0)
+			string(SUBSTRING ${tag_desc} 6 -1 tag_desc)
+	    endif()
+	else()
+	   set(tag_desc "")
+	endif()
+
+
+	set(${result} ${tag_desc} PARENT_SCOPE)
+  else(GIT_FOUND)
+	set(${result} "" PARENT_SCOPE)
```

Comment:
Use `"noversion"` or `"source"` instead of empty string?

Reply:
Empty string is equivalent to a false value in CMake. `NOTFOUND` or something suffixed with `-NOTFOUND` like `GITVERSION-NOTFOUND` should make it work the same for existing comparisons.

Reply:
I was suggesting setting the result as `tarball` or `built-from-source` to indicate that it was built from the source tarball.

Reply:
If it was built from the source tarball then that implies they got it from a tagged release and the version number should be accurate. I think most versions built by package managers (spack, vcpkg, conda-forge) would then end up with a version string of "2.0.0-noversion" or "2.0.0-source" since they'll likely be pulling the all submodules tarball for building.

Reply:
The intention was for there to be no additional string if the build was a tagged version of HELICS, so I was thinking empty, I suppose NOTFOUND would be equivalent, but I am kind of inclined to leave it empty.  Over time if the cmake function proves useful we might think about adding more options to it and making a little more flexible.